### DISCUSSION TITLE
"Missing GL version" error

### DISCUSSION THREAD

**Discussion by lbosio93** [FLAMEGPU/FLAMEGPU2#1063] [Q&A] [answered]
Hello everyone,

as Paul suggested I'm opening a discussion here to explain our problem running FLAMEGPU. 
I'm adacting a mail I previously sent to him.

Our setup is Ubuntu 20.04 (we also gave a try to Rocky 9) running in an LXD virtual machine with an NVIDIA T4 GPU, Driver Version: 530.30.02, CUDA Version: 12.1, accessed through Apache Guacamole via xrdp (yeah, we know, the typical use case). We tried running FLAMEGPU there with no luck.
We compiled the software with the following script:

 ```
#!/bin/bash
mkdir -p build && cd build
cmake .. -DCMAKE_CUDA_COMPILER=/usr/local/cuda/bin/nvcc -DCMAKE_BUILD_TYPE=Release -DCMAKE_CUDA_ARCHITECTURES=native -DFLAMEGPU_VISUALISATION=ON
cmake --build . --target all -j16
```

To run the executable /build/bin/Release/template, first we had to look at glxinfo for available visuals and exported `SDL_VIDEO_X11_VISUALID="0x022"` (we think 022 was the most suited).
We then got stuck at:
```
    Swap Interval Failed: No OpenGL context has been made current
    terminate called after throwing an instance of 'flamegpu::visualiser::GLError'
      what():  /root/cottolengo-hospital-flame-gpu-2/build/_deps/flamegpu_visualiser-src/src/flamegpu/visualiser/util/GLcheck.h(58): Glew Init failed;
    Missing GL version
```

The root cause is located in
```
build/_deps/flamegpu_visualiser-src/src/flamegpu/visualiser/Visualiser.cpp
bool Visualiser::init()
if (!this->window)    in line 677
```

and the error given by SDL_GetError() is the"Missing GL version" we printed above.

Thanks for your help.
Lorenzo

**Comment by mondus** [comment]
We have discussed this a team and have the following thoughts.

OpenGL over virtual machines is quite often problematic. There are some nuances of the virtualisation process which are specific to different hardware versions. I personally spent a lot of time messing around with such things in moving [GPU courses to the cloud](https://rse.shef.ac.uk/blog/2020-03-24-configuring-cuda-and-opengl-courses-in-the-cloud-md/).

It seems from your error that there is not an OpenGL device and the driver is reporting this with the missing GL version error. As the error comes from SDL our suggestion is to create a minimal example in the SDL repo as if you resolve this then FLAME GPU will likely work. 

If you can resolve the SDL issue then it is something which we can help with otherwise we could class this as a machine configuration issue and its out of our scope to fix/investigate.

**Comment by lbosio93** [comment]
Hello @mondus,

thanks for your answer.

As you suggested, we restricted the problem to SDL. 

We tried version 2 (the same pulled by FLAMEGPU) and version 3, and for both repositories, the [testgl](https://github.com/libsdl-org/SDL/blob/main/test/testgl.c) and all other GL tests work with no further fixes or changes from our side.

I'm attaching to the discussion a screenshot of the execution and the full text log, just in case.

Please let us know if you need other checks.
Thanks for your help.
Lorenzo

[log.txt](https://github.com/FLAMEGPU/FLAMEGPU2/files/11406621/log.txt)
![testgl](https://user-images.githubusercontent.com/47252355/236469523-95255130-9e7f-45f0-8943-3fc0ad856c10.png)

**Comment by ptheywood** [comment]
Thanks for trying the SDL2 (and 3) standalone with your setup.

Given SDL works for you on it's own, it could be one of the GL attribtues our visualiser sets which is not compatible with the OpenGL version available on your system. 

We do require OpenGL 4.3 in our visualiser (which we don't make very easy to find in the readme etc, which we should maybe address), most likely for osme Shader features we use. 

It looks like the OpenGL you have is 3.1 from the screenshot above, provided by the MESA driver, but this could just be the selected profiler, rather than the maximum provided. 

Could you run the folllowing and let us know what it reports to try and narrow down if the OpenGL version is the issue? 

```bash
glxinfo | grep "OpenGL version"
```

For me, on my local machine this returns: 

```
OpenGL version string: 4.6.0 NVIDIA 530.30.02
```

Indicating I have OpenGL 4.6 provided by the NVIDIA 530 proprietary driver.

**Comment by lbosio93** [comment, marked as answer]
Hello @ptheywood and team,

after some tests with different xorg configs we decided to move to VNC instead of RDP. 
We got the software working, with some performance loss due to protocol switch.

Thanks again for your support.
